src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl_IVsSolutionEvents.cs

```diff
@@ -59,60 +80,23 @@ int IVsSolutionEvents.OnQueryCloseSolution(object pUnkReserved, ref int pfCancel
 
         int IVsSolutionEvents.OnBeforeCloseSolution(object pUnkReserved)
         {
-            AssertIsForeground();
-
-            _solutionIsClosing = true;
-
-            foreach (var p in this.ImmutableProjects)
-            {
-                p.StopPushingToWorkspaceHosts();
-            }
-
-            _solutionLoadComplete = false;
-
-            // Cancel any background solution parsing. NOTE: This means that that work needs to
-            // check the token periodically, and whenever resuming from an "await"
-            _solutionParsingCancellationTokenSource.Cancel();
-            _solutionParsingCancellationTokenSource = new CancellationTokenSource();
-
+            DeferredState?.ProjectTracker.OnBeforeCloseSolution();
             return VSConstants.S_OK;
         }
 
         int IVsSolutionEvents.OnAfterCloseSolution(object pUnkReserved)
         {
-            AssertIsForeground();
-
-            if (IsDeferredSolutionLoadEnabled())
-            {
-                // Copy to avoid modifying the collection while enumerating
-                var loadedProjects = ImmutableProjects.ToList();
-                foreach (var p in loadedProjects)
-                {
-                    p.Disconnect();
-                }
-            }
-
-            lock (_gate)
-            {
-                Contract.ThrowIfFalse(_projectMap.Count == 0);
-            }
-
-            NotifyWorkspaceHosts(host => host.OnSolutionRemoved());
-            NotifyWorkspaceHosts(host => host.ClearSolution());
-
-            lock (_gate)
-            {
-                _projectPathToIdMap.Clear();
-            }
-
-            foreach (var workspaceHost in _workspaceHosts)
-            {
-                workspaceHost.SolutionClosed();
-            }
-
-            _solutionIsClosing = false;
-
+            DeferredState?.ProjectTracker.OnAfterCloseSolution();
             return VSConstants.S_OK;
         }
+
+        internal static bool IsDeferredSolutionLoadEnabled(IServiceProvider serviceProvider)
+        {
+            // NOTE: It is expected that the "as" will fail on Dev14, as IVsSolution7 was
```

Comment:
Note that when it's cleaned up, that should include removing the local interop definition, and adding the real reference, etc.